### PR TITLE
Add package routes

### PR DESCRIPTION
**Pull request by efriis**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -17,15 +17,14 @@ sse-starlette = {version = "^1.3.0", optional = true}
...
 langchain = ">=0.0.316"
+tomli = "^2.0.1"
 
 [tool.poetry.group.dev.dependencies]
 jupyterlab = "^3.6.1"
 fastapi = ">=0.90.1"
 sse-starlette = "^1.3.0"
 httpx-sse = ">=0.3.1"
 
-[tool.poetry.group.typing.dependencies]
```
Comment (efriis) [inline pyproject.toml:27]: Did we need this empty group for something?

File: langserve/packages.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    try:
+        module = data["tool"]["langserve"]["export_module"]
+        attr = data["tool"]["langserve"]["export_attr"]
+        package_name = data["tool"]["poetry"]["name"]
+    except KeyError as e:
+        raise KeyError("Invalid LangServe PyProject.toml") from e
+    return LangServeExport(module=module, attr=attr, package_name=package_name)
+
+
+exclude_paths = set(["__pycache__", ".venv", ".git", ".github"])
```
Comment (eyurtsev) [inline langserve/packages.py:30]: nit
```suggestion
EXCLUDE_PATHS = {"__pycache__", ".venv", ".git", ".github"}
```

File: langserve/packages.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    for part in path.parts:
+        if part in exclude_paths:
+            return False
+        if part.startswith("_"):
+            return False
+    return True
+
+
+def list_packages(path: str = "../packages") -> Generator[Path, None, None]:
```
Comment (eyurtsev) [inline langserve/packages.py:45]: Could you document how this should work?

File: langserve/packages.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    # traverse path for routes to host (any directory holding a pyproject.toml file)
+    package_root = Path(path)
+    for pyproject_path in package_root.glob("**/pyproject.toml"):
```
Comment (eyurtsev) [inline langserve/packages.py:48]: Are all packages with pyproject.toml usable or do they need `langserve` tool definition?

File: langserve/packages.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        )
+        return
+    # get attr
+    chain = getattr(mod, package["attr"])
+    add_routes(app, chain, path=mount_path)
+
+
+def add_package_routes(app: Union[FastAPI, APIRouter], path: str = "packages") -> None:
+    # traverse path for routes to host (any directory holding a pyproject.toml file)
+    for package_path in list_packages(path):
```
Comment (eyurtsev) [inline langserve/packages.py:84]: This code will break if a user creates a sub-package with pyproject.toml which is not a langserve package. Do we need to handle gracefully or do we want it to make some nosie?

File: langserve/packages.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+from fastapi import APIRouter, FastAPI
+from tomli import load
+
+from langserve.server import add_routes
+
+
+class LangServeExport(TypedDict):
+    module: str
+    attr: str
```
Comment (eyurtsev) [inline langserve/packages.py:14]: Would you mind documenting the arguments?

File: langserve/packages.py (hunk 5 above)
Comment (efriis) [reply langserve/packages.py:90]: Good call. I'll add a logging case for it to avoid throwing an error

File: langserve/packages.py (hunk 4 above)
Comment (efriis) [reply langserve/packages.py:54]: They need the `langserve` tool definition, and changed the logic to just warn if that's not the case

**Pull request merged**